### DISCUSSION TITLE
3N Integration Algorithm

### DISCUSSION THREAD

**Discussion by BraydenAlbery** [PrincetonUniversity/athena#575] [Q&A] [answered]
Hi there,

I am reading up on some background theory / numerical methods for my PhD while my simulations run.
Could someone help me out with how the 3N algorithm is implemented for ssprk5_4? I can't seem to find a paper that states the 3N algorithm explicitly. 
I am assuming the 2S algorithm for the other integrators is algorithm 3 of Ketcheson 2010 (but correct me if I am wrong :) )

$m$ = number of stages 
$S_{2}=0$ 
$S_{1} = U_{i,j,k}^{n}$
For $I=2, I\leq m+1, I= I+1$ { 
$S_{2} = S_{2}+\delta_{I-1}S_{1}$
$S_{1} = \gamma_{I,1} S_{1}+\gamma_{I,2} S_{2} + \beta_{I,I-1} \Delta t \vec{\nabla}\cdot[F_{i,j,k}^{(S_{1})}]$
} End for
$U_{i,j,k}^{n+1} = S_{1}$

Thanks for any help!

**Comment by c-white** [comment]
Apologies if I'm repeating what you already know. The coefficients are set in `src/task_list/time_integrator.cpp` (search for `ssprk5_4`). The reference there is for [Gottlieb (2009)](https://doi.org/10.1007/s10915-008-9239-z), the equation after 3.3. Since it's a paywalled article, here's the relevant equation:

<img width="614" alt="ssprk5_4" src="https://github.com/PrincetonUniversity/athena/assets/2496675/25034170-dc22-4e66-8dc9-df6b9a372b19">

Is this what you're looking for?

**Reply by BraydenAlbery** [reply]
Sorry, but not quite. I have the coefficients from time_integrator and access to that paper for the equations. I'm specifically looking for the 3N low-storage algorithm. Thank you for getting back to me!

**Comment by BraydenAlbery** [comment, marked as answer]
I think I have figured it out! It appears to be a modified version of the 3S* algorithm from Ketcheson (2010). Removing the error calculation and adding an extra term to S2 when I=5. This seems to produce the equations from Gottlieb (2009) exactly:

$m = 5$  (number of stages)
$S_{3}=U_{i,j,k}^{n}$
$S_{2}=0$
$S_{1} = U_{i,j,k}^{n}$
For $I=2,~I\leq m+1,~I= I+1$ {
$S_{2} = S_{2}+\delta_{I-1}S_{1}$
If $I==5$ { 
$S_{2} = S_{2} + \beta_{extra} \Delta t \cdot \vec{\nabla}[F_{i,j,k}(S_{1})]$
} End if
$S_{1} = \gamma_{I,1}S_{1}+\gamma_{I,2}S_{2}+\gamma_{I,3}S_{3}+\beta_{I,I-1}\Delta t \vec{\nabla}\cdot[F_{i,j,k}(S_{1})]$
} End for
$U_{i,j,k}^{n+1}=S_{1}$

**Reply by felker** [reply]
yes, I implemented this and modified it for the final stage to fit our framework implementation
